### PR TITLE
nLargest, nSmallest

### PR DESCRIPTION
**Pull request by sithmel**
https://github.com/sithmel/iter-tools/issues/39

### CODE REVIEW

**Comment by conartist6** [comment]
I'm still a bit iffy on this. I don't love the `n` in the method names. That said I see the purpose it serves from a readability perspective of documenting what the numeric parameter specifies, and I can't think of a better way to do that off the top of my head.

I'm also don't think it makes sense to put what in my mind are heap instance methods into iter-tools. If you want static versions of them in order to avoid causing the user to explicity call the heap constructor why not just make them static exports from little-ds?

**Comment by conartist6** [comment]
Ah, I had forgotten about that property of heaps. By static exports I just mean that you would use them like this:
```js
import {map} from 'iter-tools'
import {takeNSmallest} from 'little-ds-toolkit'
takeNSmallest(3, map((n, i) => n + i, [7, 2, 9, 1, 4, 12]))
```

**Comment by conartist6** [comment]
I don't know, maybe it's fine. `little-ds-toolkit` is a lot harder to remember, and it's more difficult to promote usage of two libraries

File: src/internal/take-sorted.mjs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import ensureIterable from './ensure-iterable'
+import Heap from 'little-ds-toolkit/lib/heap'
+import _iteratee from 'lodash/iteratee'
+
+export default function * takeSorted (number, expr, comparator, iterable) {
```
[the remark this answers is not in the document]
Comment (sithmel) [reply src/internal/take-sorted.mjs]: I am not sure I understand. I guess I can replace that with a boolean and import the comparator functions here, if this is what you mean

File: src/index.d.ts [hunk 2]
```diff
@@ -55,6 +55,12 @@ export declare function iterable<T>(iterable: IterableLike<T>): Iterable<T>;
 export declare function map<T, O>(func: (item: T) => O): (iter: IterableLike<T>) => Iterable<O>;
 export declare function map<T, O>(func: (item: T) => O, iter: IterableLike<T>): Iterable<O>;
 
+export declare function nLargest<T, O>(n: Number, func: (item: T) => O): (iter: IterableLike<T>) => Iterable<O>;
```
[the remark this answers is not in the document]
Comment (sithmel) [reply src/index.d.ts]: I didn't think about localeCompare. Good point.
What about exposing only the lower-level takeSorted then?
```js
takeSorted (number, [comparator,] iterable)
```

**Comment by sithmel** [comment]
I'd like to close/merge this. What about my proposal of having a single takeSorted:
```js
takeSorted (number, [comparator,] iterable)
```
@conartist6  This would remove your concerns around the iteratee function.

**Comment by conartist6** [comment]
That `takeSorted` is fine with me. I just really don't like `lodash.iteratee` there.

**Comment by sithmel** [comment]
pushing this to master directly

**Pull request closed without merging**